drivers/src/pmu.rs

```diff
@@ -50,19 +53,62 @@ impl PmuInfo {
         Ok(ControlFlow::Continue(()))
     }
 
+    fn init_delegated_counters(&mut self) -> SbiResult<()> {
+        use riscv_regs::CSR;
+        // CYCLE and INSRET must be delegated
+        const MANDATORY_DELEGATED_COUNTERS_MASK: u64 = 0x5;
+        let delegated_counters = CSR.smcounteren.get();
+        if (delegated_counters & MANDATORY_DELEGATED_COUNTERS_MASK) == 0 {
+            return Err(SbiError::NotSupported);
+        }
+
+        self.valid_hw_counters_mask = delegated_counters;
+        for i in 0..(MAX_HARDWARE_COUNTERS as u64) {
+            if (delegated_counters & (1 << i)) != 0 {
+                let base_csr = riscv_regs::CSR_CYCLE as u64;
+                let info = PmuCounterInfo::from_csr(base_csr + i);
+                self.hw_counters_info.push(Some(info));
+            } else {
+                // Hardware counters can be sparse, and may not be implemented.
+                self.hw_counters_info.push(None);
+            }
+        }
+
+        Ok(())
+    }
+
     /// Initializes the global PmuInfo structure with the information obtained from the platform SBI.
     pub fn init() -> SbiResult<()> {
         let mut pmu_info = PmuInfo::new();
-        let num_counters = pmu::get_num_counters()?;
-        for i in 0..num_counters.min(MAX_HARDWARE_COUNTERS as u64) {
-            if matches!(pmu_info.init_counter(i)?, ControlFlow::Break(())) {
-                break;
+        let cpu_id = CpuInfo::get();
+        if !cpu_id.has_sscdeleg() {
+            let num_counters = pmu::get_num_counters()?;
+            for i in 0..num_counters.min(MAX_HARDWARE_COUNTERS as u64) {
+                if matches!(pmu_info.init_sbi_counter(i)?, ControlFlow::Break(())) {
+                    break;
+                }
             }
+        } else {
+            pmu_info.init_delegated_counters()?;
         }
+
         if !pmu_info.hw_counters_info.is_empty() {
             // Stop and reset all HW counters at the outset
             let stop_flags = PmuCounterStopFlags::default().set_reset_flag();
-            let _ = pmu::stop_counters(0, pmu_info.valid_hw_counters_mask, stop_flags);
+            if !cpu_id.has_sscdeleg() {
+                let _ = pmu::stop_counters(0, pmu_info.valid_hw_counters_mask, stop_flags);
+            } else {
+                let delegated_counters = pmu_info.valid_hw_counters_mask;
+                let base_csr = riscv_regs::CSR_CYCLE as u64;
+                riscv_regs::CSR.scountinhibit.set(delegated_counters);
+                for i in 0..MAX_HARDWARE_COUNTERS as u64 {
+                    if let Ok(info) = pmu_info.get_counter_info(i) {
+                        let csr_index = info.get_csr() - base_csr;
```

Comment:
Aren't these identity-mapped, i.e. counter `i` is at CSR `cycle + i`? In which case, can't you write `CSR.scounterselect.set(i)`?

Reply:
Yup. For delegated it is identity mapped.